crates/iceberg-catalog/src/catalog/tables.rs

```diff
@@ -376,43 +377,107 @@ impl<C: Catalog, A: Authorizer + Clone, S: SecretStore>
+            // Check if table exists
+            previous_table_id = C::table_to_id(
+                warehouse_id,
+                &table,
+                ListFlags {
+                    include_active: true,
+                    include_staged: true,
+                    include_deleted: false,
+                },
+                t_read.transaction(),
+            )
+            .await?;
+
+            if let Some(previous_table_id) = previous_table_id {
+                tracing::debug!(
+                    "Register Table: Dropping existing table '{}' in namespace '{:?}' with id {previous_table_id} for overwrite operation",
+                    table.name, table.namespace
+                );
+                // Verify authorization to drop the table first
+                authorizer
+                    .require_table_action(
+                        &request_metadata,
+                        Ok(Some(previous_table_id)),
+                        CatalogTableAction::CanDrop,
+                    )
+                    .await?;
+
+                // Drop the existing table to overwrite it
+                let _previous_table_location =
+                    C::drop_table(previous_table_id, false, t_write.transaction()).await?;
+                // We don't drop the files for the previous table on overwrite
+            }
+        }
+        t_read.commit().await?;
+
         validate_table_properties(table_metadata.properties().keys())?;
         storage_profile.require_allowed_location(&table_location)?;
 
-        let namespace = C::get_namespace(warehouse_id, namespace_id, t.transaction()).await?;
         let tabular_id = TableIdentUuid::from(table_metadata.uuid());
 
         let CreateTableResponse {
             table_metadata,
             staged_table_id,
         } = C::create_table(
             TableCreation {
-                namespace_id: namespace.namespace_id,
+                namespace_id,
                 table_ident: &table,
                 table_metadata,
                 metadata_location: Some(&metadata_location),
             },
-            t.transaction(),
+            t_write.transaction(),
         )
         .await?;
 
         let config = storage_profile
             .generate_table_config(
-                DataAccess {
-                    vended_credentials: false,
-                    remote_signing: false,
-                },
+                DataAccess::not_specified(),
                 storage_secret.as_ref(),
                 &table_location,
                 StoragePermissions::ReadWriteDelete,
             )
             .await?;
 
-        authorizer
-            .create_table(&request_metadata, tabular_id, namespace_id)
-            .await?;
+        let mut auth_needs_delete = false;
+        // Delete the previous table from authorizer if it exists and differs from the new one
+        if let Some(previous_table_id) = previous_table_id {
+            if previous_table_id != tabular_id {
```

Comment:
maybe I've missed that detail - how can we be sure that identical ids are in fact the same table?